### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by paulthomasjerome**
Logic is almost working

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -78,41 +56,192 @@ const discPlacement = player => {
...
+    });
+    row += direction.row; 
+    col += direction.col;
+  }
+
+  return flipPositions;
+}
+
 /**
  * Initializes the initial state of the game and renders the board based on that state
```
Comment (beaulm) [inline index.js]: Is this comment still accurate?

> Comments are just lies waiting to happen

File: index.js [hunk 2]
```diff
@@ -78,41 +56,192 @@ const discPlacement = player => {
...
-    [' ',' ',' ',' ',' ',' ',' ',' '],
-    [' ',' ',' ','w','b',' ',' ',' '],
-    [' ',' ',' ','b','w',' ',' ',' '],
-    [' ',' ',' ',' ',' ',' ',' ',' '],
-    [' ',' ',' ',' ',' ',' ',' ',' '],
-    [' ',' ',' ',' ',' ',' ',' ',' '],    
-  ];
-
-  drawBoard(board);
+const Game = (moveRow, moveCol, player) => {
```
Comment (beaulm) [inline index.js]: A couple of comments on the variable naming here:

1. "Game" doesn't seem to describe this variable very well.
1. Unless you're defining a class, I wouldn't start with a capital letter. Constants can use ALL_CAPS but I think camelCase is probably fine for this one.

```suggestion
const makeMove = (moveRow, moveCol, player) => {
```

File: index.js [hunk 3]
```diff
@@ -78,41 +56,192 @@ const discPlacement = player => {
...
-  //for each row of the board
-    //for each space in this row
-      //if a user clicks on this space
-        //if this space is empty
-          //if this is player black
+  if(player === 'b') {
+    opponent = 'w';
+  } else {
+    opponent = 'b';
+  }
```
Comment (beaulm) [inline index.js]: This is totally fine -- it's verbose and I like that. Some other ways you could do it though:
```suggestion
  let opponent = 'b';
  if(player === 'b') {
    opponent = 'w';
  }
```

or 

```suggestion
  const opponent = (player === 'b') ? 'w' : 'b';
```

File: index.js [hunk 4]
```diff
@@ -78,41 +56,192 @@ const discPlacement = player => {
...
-  drawBoard(board);
+const Game = (moveRow, moveCol, player) => {
 
-  //for each row of the board
-    //for each space in this row
-      //if a user clicks on this space
-        //if this space is empty
-          //if this is player black
+  if(player === 'b') {
+    opponent = 'w';
```
Comment (beaulm) [inline index.js]: We should talk about variable scoping in javascript next time we hang out. [This is a big no-no](https://stackoverflow.com/questions/6888570/declaring-variables-without-var-keyword). You should _always_ use a keyword when defining variables. I'm guessing you did it this way because it worked, but there can be serious consequences to implicit globals (collisions, memory leaks etc).

TL;DR: In ES6, all variables defined with `let` or `const` are confined to the curly braces that surround them. In this case that would be the `if` statement. You should define `opponent` in an outer scope.

File: index.js [hunk 5]
```diff
@@ -78,41 +56,192 @@ const discPlacement = player => {
   });
 }
 
+const piecesToFlip = (startRow, startCol, direction, player) => {
+  //instantiate flipPositions
+  const flipPositions = [];
```
Comment (beaulm) [inline index.js]: Why are you recording the pieces to be flipped instead of just flipping them in place?

File: index.js [hunk 6]
```diff
@@ -78,41 +56,192 @@ const discPlacement = player => {
...
+  //instantiate flipPositions
+  const flipPositions = [];
+  let row = startRow;
+  let col = startCol;
+  //while we have not seen the players color
+  while(board[row][col] !== player) {
+    //if we see a blank space
+    if(board[row][col] === ' ') {
+      //dont record any positions
+      return [];
```
Comment (beaulm) [inline index.js]: Since this is in a `while` statement, won't we _always_ either hit a blank space or the edge of the board eventually?

File: index.js [hunk 7]
```diff
@@ -78,41 +56,192 @@ const discPlacement = player => {
...
+    }
+    //record the current position for our output
+    flipPositions.push({
+      row: row,
+      col: col
+    });
+    row += direction.row;
```
Comment (beaulm) [inline index.js]: What happens at the edges of the board?

File: index.js [hunk 8]
```diff
@@ -78,41 +56,192 @@ const discPlacement = player => {
...
+  console.log('the opposing player is currently ' + opponent);
+      // console.log(piecesToFlip(row, col - 1, direction));
 
-}
+  // DELCOM we should find a way simply pass the valid directions
+  // to a single valid move  html
 
-//flip over pieces depending on the board state
-const updateState = () => {
+  const direction = {
```
Comment (beaulm) [inline index.js]: This variable name could be a bit more descriptive -- it took me a while to figure out what it actually was/does.

```suggestion
  const moveBy = {
```

Eh, naming things is hard. I don't love `moveBy` either, but perhaps it's a bit better.

Perhaps we can sidestep the entire issue though: do we really need an object to contain these two variables? Why not update the method signature for `piecesToFlip` to `(startRow, startCol, horizontal, vertical, player)` and then instead of:

```js
      direction.row = 0;
      direction.col = -1;
      const flipped = piecesToFlip(row, col - 1, direction, player);
```

Just do

```js
      const flipped = piecesToFlip(row, col - 1, -1, 0, player);
```

File: index.js [hunk 9]
```diff
@@ -78,41 +56,192 @@ const discPlacement = player => {
...
+    row: 0,
+    col: 0
+  };
+    //if (0)west([row][col - 1]) adjacent is opposite color
```
Comment (beaulm) [inline index.js]: All of these `if` statements should be un-indented by one.

Also, this would technically be an **excellent** opportunity to utilize Javascripts asynchronous capabilities by flipping the stones in each direction simultaneously with [Promise.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all).

File: index.js [hunk 10]
```diff
@@ -78,41 +56,192 @@ const discPlacement = player => {
...
+      direction.col = -1;
+      const flipped = piecesToFlip(row + 1, col - 1, direction, player);
+      console.log(flipped);
+      if(flipped.length) {
+        piecesFlipped = true;
+        for(let i = 0; i < flipped.length; i++) {
+          board[flipped[i].row][flipped[i].col] = player;
+        }
+      }
+    }
```
Comment (beaulm) [inline index.js]: Hmmm, seems awfully WET to me...

File: index.js [hunk 11]
```diff
@@ -78,41 +56,192 @@ const discPlacement = player => {
...
 }
 
-startGame();
-//DELCOM we may have to encapsulate this in its own function to be part of the board drawing method
-discPlacement(player);
+//instantiate the first player's color
+let currentPlayer = 'b';
+
+//set the initial state of the logical board
+const board = [
```
Comment (beaulm) [inline index.js]: We should also talk about ["hoisting"](https://developer.mozilla.org/en-US/docs/Glossary/Hoisting). You're relying on this being hoisted and available in the Global scope, which is generally quite discouraged. I think you'd be much better off passing the board directly into your `Game` function and then getting it back, which would make `Game` a "pure" function which has many advantages, the least of which is _much_ easier testability.

...speaking of which, this would be an _excellent_ point in your project to add tests and start doing TDD. Happy to help you out with that part if you want.

File: index.js [hunk 12]
```diff
@@ -78,41 +56,192 @@ const discPlacement = player => {
...
+  [' ',' ',' ',' ',' ',' ',' ',' '],
+  [' ',' ',' ',' ',' ',' ',' ',' '],
+  [' ',' ',' ',' ',' ',' ',' ',' '],
+  [' ',' ',' ','w','b',' ',' ',' '],
```
Comment (beaulm) [inline index.js]: Totally fine, but why did you use strings instead of `null`,`0`,`1` here?

**Review by beaulm — COMMENTED** [review]
Despite all the comments, this actually looks great! You're definitely on the right track. A few improvements will make the code MUCH easier to read/understand/update, and the rest will hopefully slide right into place after that.

File: index.js (hunk 1 above)
Comment (paulthomasjerome) [reply index.js]: No i just updated it though, thanks!

File: index.js (hunk 2 above)
Comment (paulthomasjerome) [reply index.js]: Awesome, that is a change i want to make, this function has evolved alot in the past 12 hours (as you could tell from the old comment) I will make that change now

File: index.js (hunk 3 above)
Comment (paulthomasjerome) [reply index.js]: I am starting to really like ternary ops (oddly enough lol) so i will implement that

Comment (beaulm) [reply index.js]: Yeah, it took me a _looooong_ time but I finally came around on them. Super effin' pumped for Optional Chaining and Nullish Coalescing which [just got accepted into the spec!](https://2ality.com/2019/12/ecmascript-2020.html)

File: index.js (hunk 4 above)
Comment (paulthomasjerome) [reply index.js]: Moved it to the outer most scope where the board and initial player are instantiated

File: index.js (hunk 5 above)
Comment (paulthomasjerome) [reply index.js]: I suppose i could do that and my logic to check if pieces were flipped should still work as long as I still return something to check?

File: index.js (hunk 10 above)
Comment (paulthomasjerome) [reply index.js]: Once i encapsulate the flipping within `const flipped` would it be more DRY?

File: index.js (hunk 11 above)
Comment (paulthomasjerome) [reply index.js]: I really want to do that and would really appreciate the help in doing so

File: index.js (hunk 12 above)
Comment (paulthomasjerome) [reply index.js]: I suppose it was just for conceptual purposes, it would use less resources to use `null`,`0`,`1` correct?

File: index.js (hunk 6 above)
Comment (paulthomasjerome) [reply index.js]: Hmmm... I am not sure what else to do here. The reason I have done this is to try to indicate that no pieces should be flipped. I need to think about it a bit

File: index.js (hunk 5 above)
Comment (beaulm) [reply index.js]: Yup!

File: index.js (hunk 6 above)
Comment (beaulm) [reply index.js]: Once you start flipping in place you'll probably figure out what to do here

File: index.js (hunk 10 above)
Comment (beaulm) [reply index.js]: Yeah, but you probably don't even need 7 or 8 hard-coded blocks to accomplish what all this is doing

File: index.js (hunk 11 above)
Comment (beaulm) [reply index.js]: You got it!

File: index.js (hunk 12 above)
Comment (beaulm) [reply index.js]: Eh, probably, but not by any meaningful amount. More just a programmer thing. I'll have to think about it a bit more.

File: index.js (hunk 7 above)
Comment (paulthomasjerome) [reply index.js]: That is something i still need to figure out ;p I also need to figure out what happens in the cases where a player cant make a valid move due to it being impossible to flip pieces

File: index.js (hunk 3 above)
Comment (paulthomasjerome) [reply index.js]: I was just reading about ES2020 stuff randomly this morning! We should talk about that more tonight!

File: index.js (hunk 5 above)
Comment (paulthomasjerome) [reply index.js]: done

File: index.js (hunk 8 above)
Comment (paulthomasjerome) [reply index.js]: done

File: index.js (hunk 6 above)
Comment (paulthomasjerome) [reply index.js]: Ended up getting rid of these lines

File: index.js (hunk 9 above)
Comment (paulthomasjerome) [reply index.js]: yeah they didnt paste properly lol i fixed that, still need to look into Promise.all

File: index.js (hunk 10 above)
Comment (paulthomasjerome) [reply index.js]: Yeah i need to figure that out

File: index.js (hunk 12 above)
Comment (paulthomasjerome) [reply index.js]: Ok well i did update it to be null, 0, 1

File: index.js (hunk 10 above)
Comment (beaulm) [reply index.js]: I wonder if you could do something like this:
```js
  for(let vertical = -1; vertical <= 1; vertical++) {
    for(let horizontal = -1; horizontal <= 1; horizontal++) {

    }
  }
```

Comment (beaulm) [reply index.js]: Just looked at your latest code and updated to `vertical`/`horizontal`

File: index.js [hunk 13]
```diff
@@ -1,118 +1,171 @@
...
+
+  //instantiate vertical and horizontal translation
+  let vertical = 0;
+  let horizontal = 0;
+
+  //if (0)west([row][col - 1]) adjacent is opposite color
+  if(board[row][col - 1] === opponent) {
+    vertical = 0;
+    horizontal = -1;
+    const flipped = piecesToFlip(row, col - 1, vertical, horizontal, player);
```
Comment (beaulm) [inline index.js]:
```suggestion
    const flipped = piecesToFlip(row + vertical, col + horizontal, vertical, horizontal, player);
```

File: index.js [hunk 14]
```diff
@@ -1,118 +1,171 @@
...
+      piecesFlipped = true;
+    }
+  }     
+
+  //if any pieces were flipped
+  if(piecesFlipped) {
+    //this move is valid and the piece can be placed where the player has chosen
+    board[row][col] = player;
+    //change players
+    currentPlayer = opponent;
```
Comment (beaulm) [inline index.js:55]: You technically don't even need to store the value of the opponent:

```suggestion
    currentPlayer = !currentPlayer;
```

Comment (paulthomasjerome) [reply index.js:55]: I don't really understand that, can you tell me why?

Comment (beaulm) [reply index.js:55]: Because you're storing the `currentPlayer` as a boolean now, so all you have to do is flip the bit.

I guess this is one of those advantages of using `0`/`1` instead of strings.

Comment (beaulm) [reply index.js:55]: And you don't need to store the value of the opponent since computers are so good at binary manipulation that you could easily just use `!currentPlayer` anywhere you would use `opponent` (though it's fine to keep `opponent` if you think it's more clear).

File: index.js [hunk 15]
```diff
@@ -1,118 +1,124 @@
...
+  let piecesFlipped = false;
+
+  //initialize flag to track if we have made a valid move
+  let validMove = false;
+
+  //check all directions
+  for (let vertical = -1; vertical <= 1; vertical++) {
+    for (let horizontal = -1; horizontal <= 1; horizontal++) {
+      //if the adjacent piece is the opponents piece
+      if (board[moveRow + vertical][moveCol + horizontal] === opponent) {
```
Comment (beaulm) [inline index.js:38]: I feel like this check shouldn't be necessary since `flipPieces` will handle this logic internally as well.

(It might be necessary the way you currently have the code, but it _shouldn't_ be necessary).

**Review by beaulm — COMMENTED** [review]
Looks fantastic man! Perhaps the next thing to do would be to split `processMove` and `flipPieces` out into their own separate files and then import them into `index.js`. Doing so will help keep your code organized, reduce your reliance on global variables, and make it easier to test things in the future. Though I'm not sure if you'd need a build tool then...

File: index.js (hunk 14 above)
Comment (beaulm) [reply index.js:55]: Derp, my bad. `currentPlayer`/`opponent` are _booleans_ which aren't technically the same thing as binary. So if `currentPlayer === 0` then `currentPlayer = !currentPlayer;` is really saying `currentPlayer = !0;` and `!0` is `true` not `1` since the esclamation point coerces to a boolean, then does the "not" 🤦 

You could technically use the [unary + operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Arithmetic_Operators#Unary_plus) to convert the boolean back to binary like: `currentPlayer = +!currentPlayer;` but since that's a pretty obscure operator (I had to look it up), using the ternary operator may be more maintainable. Totally up to you!
